### PR TITLE
global face auth, account linking using tokens

### PR DESCRIPTION
**Pull request by ice-cronus**

### CODE REVIEW

File: cmd/eskimo-hut/eskimo_hut.go [hunk 1]
```diff
@@ -67,7 +68,8 @@ func (s *service) Init(ctx context.Context, cancel context.CancelFunc) {
 	s.tokenRefresher = auth.NewRefresher(server.Auth(ctx), s.authEmailLinkClient, s.telegramAuthClient)
 	s.socialRepository = social.New(ctx, s.usersProcessor)
 	s.quizRepository = kycquiz.NewRepository(ctx, s.usersProcessor)
-	s.faceKycClient = facekyc.New(ctx, s.usersProcessor)
+	s.usersLinker = linkerkyc.NewAccountLinker(ctx)
+	s.faceKycClient = facekyc.New(ctx, s.usersProcessor, s.usersLinker)
```
Comment (ice-ares) [inline cmd/eskimo-hut/eskimo_hut.go]: add them in the close section

File: application.yaml [hunk 2]
```diff
@@ -65,12 +75,18 @@ kyc/social:
...
     bafToken: bogus
     bafHost: http://bogus.com/
     availabilityUrl:
     secretApiToken: bogus-secret
     concurrentUsers: 1
+kyc/linking:
+  tenant: sunwaves
+  tenantURLs:
+    callfluent: https://localhost:1444/
+    doctorx: https://localhost:1445/
```
Comment (ice-ares) [inline application.yaml]: u can build that dynamically in the code; api.ice.io/$tenant/..

**Review by ice-ares — DISMISSED** [review]

File: kyc/face/face.go [hunk 3]
```diff
@@ -44,22 +51,40 @@ func (c *client) CheckStatus(ctx context.Context, user *users.User, nextKYCStep
...
 		}
+		if verifiedTenantFrom3rdParty != "" {
+			verifiedTenant = verifiedTenantFrom3rdParty
+		}
 	}
 	if hasResult {
-		if dErr := c.deleteUserForwarded(ctx, user.ID); dErr != nil {
-			return false, errors.Wrapf(err, "failed to delete user forwarded to face kyc for user id %v", user.ID)
+		if dErr := c.accountsLinker.SetTenantVerified(ctx, user.ID, verifiedTenant); dErr != nil {
+			return false, errors.Wrapf(err, "failed to save user face id kyc for user id %v", user.ID)
```
Comment (ice-myles) [inline kyc/face/face.go]:
```
+return false, errors.Wrapf(dErr, "failed to save user face id kyc for user id %v", user.ID)
```

File: kyc/face/face.go [hunk 4]
```diff
@@ -44,22 +51,40 @@ func (c *client) CheckStatus(ctx context.Context, user *users.User, nextKYCStep
...
+		}
+		if dErr := c.saveFaceID(ctx, user.ID, faceID); dErr != nil {
+			return false, errors.Wrapf(err, "failed to save user face id kyc for user id %v", user.ID)
```
Comment (ice-myles) [inline kyc/face/face.go]:
```
+return false, errors.Wrapf(dErr, "failed to save user face id kyc for user id %v", user.ID)
```

File: kyc/linking/linking.go [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+		Tenant       string     `db:"tenant"`
+		LinkedTenant string     `db:"linked_tenant"`
+		LinkedUserID string     `db:"linked_user_id"`
+		HasKYC       bool       `db:"has_kyc"`
+	}](ctx, l.globalDB, `WITH RECURSIVE rec AS (
+								SELECT * FROM linked_user_accounts WHERE user_id = $1 OR linked_user_id = $1
+								UNION SELECT linked_user_accounts.* FROM linked_user_accounts
+								JOIN rec on rec.user_id = linked_user_accounts.user_id OR rec.linked_user_id = linked_user_accounts.linked_user_id
+							) SELECT * FROM rec;`, userID)
+	if err != nil && storage.IsErr(err, storage.ErrNotFound) {
```
Comment (ice-myles) [inline kyc/linking/linking.go]: Typo?
```
+!storage.IsErr(err, storage.ErrNotFound)
```
?

File: kyc/linking/linking.go [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+}
+
+func (l *linker) verifyToken(ctx context.Context, userID, tenant, token string) (remoteID UserID, hasFaceResult bool, err error) {
+	usr, err := l.fetchTokenData(ctx, tenant, token)
+	if err != nil {
+		if errors.Is(err, errRemoteUserNotFound) {
+			return "", false, ErrNotOwnRemoteUser
+		}
+
+		return "", false, errors.Wrapf(err, "failed to link accounts with %v", userID)
```
Comment (ice-myles) [inline kyc/linking/linking.go]: Seems like error description should be another

File: kyc/face/face.go (hunk 3 above)
Comment (ice-myles) [reply kyc/face/face.go]: And error description

File: cmd/eskimo-hut/kyc.go [hunk 7]
```diff
@@ -313,3 +316,46 @@ func (s *service) TryResetKYCSteps( //nolint:gocritic,funlen,gocognit,revive,cyc
...
+	req *server.Request[ForwardToFaceKYCRequestBody, ForwardToFaceKYCResponse],
+) (*server.Response[ForwardToFaceKYCResponse], *server.Response[server.ErrorResponse]) {
+	ctx = facekyc.ContextWithClientType(ctx, req.Data.XClientType) //nolint:revive // .
+	kycFaceAvailable, err := s.faceKycClient.ForwardToKYC(ctx, req.Data.UserID, req.Data.Tokens)
+	if err != nil {
+		switch {
+		case errors.Is(err, linking.ErrNotOwnRemoteUser):
+			return nil, server.BadRequest(err, linkingNotOwnedProfile)
+		case errors.Is(err, linking.ErrDuplicate):
+			return nil, server.BadRequest(err, linkingDuplicate)
```
Comment (ice-myles) [inline cmd/eskimo-hut/kyc.go]: Maybe server.Conflict here?

**Review by ice-ares — DISMISSED** [review]

**Review by ice-myles — DISMISSED** [review]

File: application.yaml [hunk 8]
```diff
@@ -30,8 +30,18 @@ cmd/eskimo-hut:
     refreshExpirationTime: 24h
     accessExpirationTime: 1h
     jwtSecret: bogus
+global:
```
Comment (ice-ares) [inline application.yaml]: actually, wait; now i see what u did here; no, we don't need the global connector; use a regular one like all the other ones

**Pull request merged**